File: packages/core/src/services/query/incantation.ts
import { Incantation as IncantationData, QueryTimeToLive } from '../../types.js';
import { RecordId, Duration } from 'surrealdb';

// Helper to parse duration string like "10m" to ms
function parseDuration(duration: QueryTimeToLive | Duration): number {
  if (duration instanceof Duration) {
    // Duration in surrealdb.js (check property)
    // Coerce to number to avoid BigInt mixing issues
    // Using string conversion fallback if specific props aren't reliable
    const ms = (duration as any).milliseconds || (duration as any)._milliseconds;
    if (ms) return Number(ms);

    // Fallback: try parsing string representation
    const str = duration.toString();
    if (str !== '[object Object]') return parseDuration(str as any);

    return 600000;
  }

  if (typeof duration === 'bigint') {
    return Number(duration);
  }

  if (typeof duration !== 'string') return 600000; // fallback

  const match = duration.match(/^(\d+)([smh])$/);
  if (!match) return 600000; // default 10m
  const val = parseInt(match[1], 10);
  const unit = match[2];
  switch (unit) {
    case 's':
      return val * 1000;
    case 'h':
      return val * 3600000;
    case 'm':
    default:
      return val * 60000;
  }
}

export class Incantation<T> {
  public id: RecordId<string>;
  public surrealql: string;
  public params?: Record<string, any>;
  public localHash: string;
  public localTree: any;
  public remoteHash: string;
  public remoteTree: any;
  public ttl: QueryTimeToLive | Duration;
  public lastActiveAt: Date | number | string;
  private ttlTimer: NodeJS.Timeout | null = null;
  private ttlDurationMs: number;
  private results: T[] | null = null;
  private meta: IncantationData['meta'];

  get records() {
    return this.results;
  }

  get tableName() {
    return this.meta.tableName;
  }

  constructor(data: IncantationData) {
    this.id = data.id;
    this.surrealql = data.surrealql;
    this.params = data.params;
    this.localHash = data.localHash;
    this.localTree = data.localTree;
    this.remoteHash = data.remoteHash;
    this.remoteTree = data.remoteTree;
    this.lastActiveAt = new Date(data.lastActiveAt);
    this.ttl = data.ttl;
    this.ttlDurationMs = parseDuration(data.ttl);
    this.meta = data.meta;
  }

  public invlovesTable(tableName: string) {
    if (this.tableName === tableName) return true;
    return this.meta.involvedTables?.includes(tableName) ?? false;
  }

  public updateLocalState(records: T[], localHash: string, localTree: any) {
    this.results = records;
    this.localHash = localHash;
    this.localTree = localTree;
  }

  public destroy() {
    this.stopTTLHeartbeat();
  }

  public startTTLHeartbeat(onHeartbeat: () => void) {
    if (this.ttlTimer) return;

    // Schedule next check.
    // Requirement: "call 10% before the TTL exceeds" => 90% of TTL.
    const heartbeatTime = Math.floor(this.ttlDurationMs * 0.9);

    // Ensure we don't spam if TTL is very short, but for "10m" (600s), 90% is 540s.
    this.ttlTimer = setTimeout(() => {
      onHeartbeat();
      this.startTTLHeartbeat(onHeartbeat);
    }, heartbeatTime);
  }

  private stopTTLHeartbeat() {
    if (this.ttlTimer) {
      clearTimeout(this.ttlTimer);
      this.ttlTimer = null;
    }
  }
}

File: packages/flutter_core/lib/core/services/query/incantation.dart
import 'dart:async';
import '../../types.dart'; // Importiert dein Enum und die Extension

/// Hilfsklasse für die Metadaten (in TS war das ein inline object type)
class IncantationMeta {
  final String tableName;
  IncantationMeta({required this.tableName});
}

class Incantation<T> {
  // Wir nutzen String für RecordId, bis du das SurrealDB Paket einbindest
  final String id;
  final String surrealql;
  final Map<String, dynamic>? params;

  // Mutable Properties, da sie sich via updateLocalState ändern
  String hash;
  dynamic tree; // Entspricht 'any' in TS, später 'IdTree'

  final QueryTimeToLive ttl;
  DateTime lastActiveAt;

  final IncantationMeta meta;

  // Private Felder für internen State
  Timer? _ttlTimer;
  late final int _ttlDurationMs;
  List<T>? _results;

  // Getters
  List<T>? get records => _results;
  String get tableName => meta.tableName;

  Incantation({
    required this.id,
    required this.surrealql,
    required this.hash,
    required this.ttl,
    required this.meta,
    this.params,
    this.tree,
    DateTime? lastActiveAt,
  }) : lastActiveAt = lastActiveAt ?? DateTime.now() {
    // Initialisiere die Dauer direkt im Konstruktor
    _ttlDurationMs = _parseDuration(ttl);
  }

  /// Aktualisiert den lokalen State mit neuen Daten vom Server oder Cache
  void updateLocalState(List<T> records, String hash, dynamic tree) {
    _results = records;
    this.hash = hash;
    this.tree = tree;
  }

  /// Stoppt den Heartbeat und räumt auf
  void destroy() {
    _stopTTLHeartbeat();
  }

  /// Startet den rekursiven Heartbeat Timer
  void startTTLHeartbeat(void Function() onHeartbeat) {
    if (_ttlTimer != null && _ttlTimer!.isActive) return;

    // Requirement: "call 10% before the TTL exceeds" => 90% of TTL.
    // .floor() rundet ab, um eine saubere Integer-Millisekunden-Zahl zu erhalten.
    final heartbeatTimeMs = (_ttlDurationMs * 0.9).floor();

    _ttlTimer = Timer(Duration(milliseconds: heartbeatTimeMs), () {
      onHeartbeat();
      // Rekursiver Aufruf, wie im TS Code
      // Wichtig: Timer ist one-shot, daher müssen wir ihn neu starten
      _stopTTLHeartbeat();
      startTTLHeartbeat(onHeartbeat);
    });
  }

  void _stopTTLHeartbeat() {
    _ttlTimer?.cancel();
    _ttlTimer = null;
  }

  /// Hilfsfunktion: Konvertiert TTL Enum in Millisekunden
  /// Portiert die Regex-Logik von `parseDuration` aus der TS-Datei.
  int _parseDuration(QueryTimeToLive ttl) {
    // Wir holen den String-Wert aus deiner Extension (z.B. "10m")
    final durationString = ttl.value;

    final regex = RegExp(r'^(\d+)([smh])$');
    final match = regex.firstMatch(durationString);

    if (match == null) {
      return 600000; // default 10m
    }

    final val = int.parse(match.group(1)!);
    final unit = match.group(2);

    switch (unit) {
      case 's':
        return val * 1000;
      case 'h':
        return val * 3600000;
      case 'm':
      default:
        return val * 60000;
    }
  }
}
